text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,262 @@
+
+```rust
+pub struct Scope<'a> {
+    dtors: RefCell<Option<DtorChain<'a>>>
+}
+
+struct DtorChain<'a> {
+    dtor: Box<FnBox() + 'a>,
+    next: Option<Box<DtorChain<'a>>>
+}
+
+pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R {
+    let mut scope = Scope { dtors: RefCell::new(None) };
+    let ret = f(&scope);
+    scope.drop_all();
+    ret
+}
+
+impl<'a> Scope<'a> {
+    // This method is carefully written in a transactional style, so
+    // that it can be called directly and, if any dtor panics, can be
+    // resumed in the unwinding this causes. By initially running the
+    // method outside of any destructor, we avoid any leakage problems
+    // due to #14875.
+    fn drop_all(&mut self) {
+        loop {
+            // use a separate scope to ensure that the RefCell borrow
+            // is relinquished before running `dtor`
+            let dtor = {
+                let mut dtors = self.dtors.borrow_mut();
+                if let Some(mut node) = dtors.take() {
+                    *dtors = node.next.take().map(|b| *b);
+                    node.dtor
+                } else {
+                    return
+                }
+            };
+            dtor()
+        }
+    }
+
+    pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a {
+        let mut dtors = self.dtors.borrow_mut();
+        *dtors = Some(DtorChain {
+            dtor: Box::new(f),
+            next: dtors.take().map(Box::new)
+        });
+    }
+}
+
+impl<'a> Drop for Scope<'a> {
+    fn drop(&mut self) {
+        self.drop_all()
+    }
+}
+```
+
+This implementation does a few interesting things:
+
+* It avoids any allocation in the case of a single deferred computation.
+
+* It works around issue 14875 in a somewhat subtle way: the `drop_all` method is
+  called *both* normally within `scope`, and in the `Scope` destructor. The
+  method is also coded transactionally. This means that the first panic (if any)
+  in a deferred computation triggers the drop for `Scoped` (without any leakage,
+  avoiding #14875), and any remaining panics wind up aborting.
+
+Note that this is just a sketch: the use of a linked list here, and `RefCell`
+internally, could both be switched for something more interesting.
+
+# Drawbacks
+
+The main drawback of this approach is that it is arguably less ergonomic (and
+less pretty) than the original `thread::scoped` API. On the flip side, as the
+RFC argues, this API makes the control-flow/synchronization of the joins much
+clearer, resolving a tension in the design of `thread::scoped` (which `must_use`
+only somewhat mitigated).
+
+# Alternatives
+
+The main alternative would be to make the original `thread::scoped` API safe by
+changing some other aspect of the language. Ideas along these lines are being
+[debated elsewhere](https://github.com/rust-lang/rfcs/pull/1066).
+
+# Unresolved questions
+
+Is there any reason that `Scope` should be `Send` or `Sync`?
```

Comment:
I can't see any reason for `Send` (we never get an owned `Scope` so we can't possibly send it anywhere) but we definitely want `Sync`, so we can pass the `Scope` to a fork-join thread in case it needs to conditionally spawn more threads (or otherwise register "finalizers" that work on shared stack data).